source/common/network/apple_dns_impl.cc

```diff
@@ -79,12 +91,20 @@ void AppleDnsResolverImpl::initializeMainSdRef() {
   // for kDNSServiceFlagsShareConnection in dns_sd.h, and copied (and edited) in this implementation
   // where relevant.
   auto error = DnsServiceSingleton::get().dnsServiceCreateConnection(&main_sd_ref_);
-  RELEASE_ASSERT(error == kDNSServiceErr_NoError,
-                 fmt::format("error={} in DNSServiceCreateConnection", error));
+  if (error != kDNSServiceErr_NoError) {
+    stats_.connection_failure_.inc();
+    initialize_failure_timer_->enableTimer(
+        std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
+    return;
+  }
 
   auto fd = DnsServiceSingleton::get().dnsServiceRefSockFD(main_sd_ref_);
-  RELEASE_ASSERT(fd != -1, "error in DNSServiceRefSockFD");
-  ENVOY_LOG(debug, "DNS resolver has fd={}", fd);
+  if (fd == -1) {
```

Comment:
I can see the above one failing, but can this one actually fail? Can you add a few more comments about what failures we are actually protecting against here?

Reply:
Yeah that call can fail according to the very opaque api docs in the apple header. Unfortunately, no good details on the failure scenario from the doc.